controllers/databaseclusterrestore_controller.go

```diff
@@ -78,32 +79,59 @@ func (r *DatabaseClusterRestoreReconciler) Reconcile(ctx context.Context, req ct
 		return reconcile.Result{}, err
 	}
 
-	if cr.Spec.DatabaseType == everestv1alpha1.DatabaseEnginePXC {
-		if err := r.restorePXC(cr); err != nil { //nolint:contextcheck
+	if (cr.Spec.DataSource.DBClusterBackupName == "" && cr.Spec.DataSource.BackupSource == nil) ||
+		(cr.Spec.DataSource.DBClusterBackupName != "" && cr.Spec.DataSource.BackupSource != nil) {
+		return reconcile.Result{}, errors.New("specify either dbClusterBackupName or backupSource")
+	}
+
+	dbCRNamespacedName := types.NamespacedName{
+		Name:      cr.Spec.DBClusterName,
+		Namespace: cr.Namespace,
+	}
+	dbCR := &everestv1alpha1.DatabaseCluster{}
+	err = r.Get(ctx, dbCRNamespacedName, dbCR)
+	if err != nil {
+		if err = client.IgnoreNotFound(err); err != nil {
+			logger.Error(err, "unable to fetch DatabaseCluster")
+		}
+		return reconcile.Result{}, err
+	}
+
+	if dbCR.Spec.Engine.Type == everestv1alpha1.DatabaseEnginePXC {
+		if err := r.restorePXC(ctx, cr); err != nil {
 			logger.Error(err, "unable to restore PXC Cluster")
 			return reconcile.Result{}, err
 		}
 	}
-	if cr.Spec.DatabaseType == everestv1alpha1.DatabaseEnginePSMDB {
-		if err := r.restorePSMDB(cr); err != nil { //nolint:contextcheck
-			logger.Error(err, "unable to restore PXC Cluster")
+	if dbCR.Spec.Engine.Type == everestv1alpha1.DatabaseEnginePSMDB {
+		if err := r.restorePSMDB(ctx, cr); err != nil {
+			// The DatabaseCluster controller is responsible for updating the
+			// upstream DB cluster with the necessary storage definition. If
+			// the storage is not defined in the upstream DB cluster CR, we
+			// requeue the backup to give the DatabaseCluster controller a
+			// chance to update the upstream DB cluster CR.
+			if errors.Is(err, ErrBackupStorageUndefined) {
+				return reconcile.Result{Requeue: true}, nil
```

Comment:
why only the psmdb gives another chance to reconcile but not the other operators?

Reply:
The other operators don't need to have the storage defined in their CRs for the restore process.
They only require it for the backups.